src/tools/rbd_mirror/LeaderWatcher.cc

```diff
@@ -0,0 +1,712 @@
+  m_timer_task = nullptr;
+}
+
+void LeaderWatcher::schedule_timer_task(const std::string &name,
+                                        int delay_factor, bool leader,
+                                        void (LeaderWatcher::*callback)()) {
+  assert(m_threads->timer_lock.is_locked());
+  assert(m_lock.is_locked());
+
+  cancel_timer_task();
+
+  m_timer_task = new FunctionContext(
+    [this, callback, leader](int r) {
+      assert(m_threads->timer_lock.is_locked());
+      m_timer_task = nullptr;
+      Mutex::Locker locker(m_lock);
+      if (is_leader(m_lock) != leader) {
+        return;
+      }
+      (this->*callback)();
+    });
+
+  int after = delay_factor *
+    max(1, m_cct->_conf->rbd_mirror_leader_heartbeat_interval);
+
+  dout(20) << "scheduling " << name << " after " << after << " sec (task "
+           << m_timer_task << ")" << dendl;
+  m_threads->timer->add_event_after(after, m_timer_task);
+}
+
+void LeaderWatcher::handle_post_acquire_leader_lock(int r, Context *on_finish) {
+  dout(20) << "r=" << r << dendl;
+
+  if (r < 0) {
+    if (r == -EAGAIN) {
+      dout(20) << "already locked" << dendl;
+    } else {
+      derr << "error acquiring leader lock: " << cpp_strerror(r) << dendl;
+    }
+    on_finish->complete(r);
+    return;
+  }
+
+  Mutex::Locker locker(m_lock);
+  assert(m_on_finish == nullptr);
+  m_on_finish = on_finish;
+  m_on_finish_r = 0;
+
+  init_status_watcher();
+}
+
+void LeaderWatcher::handle_pre_release_leader_lock(Context *on_finish) {
+  dout(20) << dendl;
+
+  Mutex::Locker locker(m_lock);
+  assert(m_on_finish == nullptr);
+  m_on_finish = on_finish;
+  m_on_finish_r = 0;
+
+  pre_release_leader();
+}
+
+void LeaderWatcher::handle_post_release_leader_lock(int r, Context *on_finish) {
+  dout(20) << "r=" << r << dendl;
+
+  Mutex::Locker locker(m_lock);
+  assert(m_on_finish == nullptr);
+  m_on_finish = on_finish;
+  m_on_finish_r = r;
+
+  notify_lock_released();
+}
+
+void LeaderWatcher::break_leader_lock() {
+  dout(20) << dendl;
+
+  assert(m_lock.is_locked());
+
+  if (m_locker.cookie.empty()) {
+    acquire_leader_lock();
+    return;
+  }
+
+  Context *ctx = create_context_callback<
+    LeaderWatcher, &LeaderWatcher::handle_break_leader_lock>(this);
+
+  m_leader_lock->break_lock(m_locker, true, async(ctx));
+}
+
+void LeaderWatcher::handle_break_leader_lock(int r) {
+  dout(20) << "r=" << r << dendl;
+
+  Mutex::Locker timer_locker(m_threads->timer_lock);
+  Mutex::Locker locker(m_lock);
+
+  if (m_leader_lock->is_shutdown()) {
+    dout(20) << "canceling due to shutdown" << dendl;
+    return;
+  }
+
+  if (r < 0 && r != -ENOENT) {
+    derr << "error beaking leader lock: " << cpp_strerror(r)  << dendl;
+
+    schedule_timer_task("get locker", 1, false, &LeaderWatcher::get_locker);
+    return;
+  }
+
+  acquire_leader_lock();
+}
+
+void LeaderWatcher::get_locker() {
+  dout(20) << dendl;
+
+  assert(m_lock.is_locked());
+
+  std::shared_ptr<librbd::managed_lock::Locker> locker(
```

Comment:
Actually, I expect concurrent get locker requests. They are generated by these events:
- lock acquire failed, 
- lock break fail,
- lock released,
- "heartbeat" message received,
- "lock acquired" message received,

as these are both internal and external events, it needs some serialization if we want to avoid concurrent requests. In my previous version I had a check in get_locker to skip if there is already an in-flight request. This requires two additional member variables (locker and a "in-flight" boolean). Using a shared pointer looked a little nicer to me, but I don't have a strong opinion here.